### DISCUSSION TITLE
Just some random thoughts

### DISCUSSION THREAD

**Discussion by Biswa96** [nmeum/android-tools#92] [Ideas]
Here, I would like to ask some random casual questions. Please answer to the question in the reply section without creating a different thread. It would help to organize the ideas.

**Comment by Biswa96** [comment]
1. Would you like to extend this project to other platforms, for example, MS Windows, BSDs, MacOS etc. ?

**Reply by anatol** [reply]
Adding multiplatform support (+win +bsd +darwin) is a great idea! If it is something you want to look at then I fully support it.

As of MacOSX there was a user request for it - https://github.com/nmeum/android-tools/issues/55. NixOS went ahead and patched the project to build it at macosx https://github.com/NixOS/nixpkgs/pull/197959/files

**Comment by Biswa96** [comment]
2. Would you like to accept patches to enable mdns option in adb?

**Reply by anatol** [reply]
How big are the patches? Do they affect other OS/configurations? But if it helps other OS then I think it is reasonable to accept it.

Have you checked with upstream what they think about the patches?

**Reply by Biswa96** [reply]
The mdns feature requires two small projects openscreen and mdnsresponder. I have not fully built it. But at first look, the required changes seem small and can be added upstream. I was waiting for green signal from here :))

**Reply by anatol** [reply]
Sounds good, then. I have no issue with small patches that are required for multi-OS support.

**Reply by Biswa96** [reply]
Enabling mdns in adb requires [openscreen ](https://android.googlesource.com/platform/external/openscreen.git) and it requires abseil library. Should abseil library be used from distribution package or from third_party directory in openscreen?

**Reply by Biswa96** [reply]
Got my own answer. Ubuntu 20.04 focal does not have the libabsl-dev package. So, I shall use the third_party/abseil for openscreen.

**Comment by Biswa96** [comment]
3. Would it be easier/faster/lighter to use tagged source tarball instead of git submodules? For example, core and base are relatively bigger repository and not all subfolders are required to compile adb. So, if the required subfolders are extracted before building, wouldn't it be faster than cloning whole repository?

**Reply by nmeum** [reply]
We are currently using shallow clones for submodules so we are not checking out the entire git repository for submodules so not sure how much faster it would really be. However, it would indeed be possible to use external tarballs via CMake's [`ExternalProject_Add`](https://cmake.org/cmake/help/latest/module/ExternalProject.html) I haven't experimented with that yet so not sure what the benefits would be.

**Reply by Biswa96** [reply]
That CMake function would extract whole source directory. Instead of using cmake for pre-build steps, I was thinking to use a shell script, like bootstrap.sh or autogen.sh does in autotools project. The steps would be - 1. shell script download, extract specific directories and patch files 2. then cmake builds those as it does now. What do you think about that?

The idea is somewhat inspired by the `prepare()` function in PKGBUILD or APKBUILD.

**Reply by nmeum** [reply]
I would prefer not to have a separate shell script and instead do everything within the CMake build system. I am also not sure if downloading tarballs is really significantly faster than git shallow submodule clones.

**Comment by Biswa96** [comment]
4. README has a line, says

> Currently the build system doesn't check whether all of these are installed.

Are there any programs or libraries which are not checked yet? In the macOS addition, most of the libraries are checked with pkgconfig.

**Reply by nmeum** [reply]
IIRC: Previously, not all dependencies had CMake `find_package` files and thus for some of them one just got build errors if they weren't installed. I haven't checked recently if that is still the case.

**Comment by Biswa96** [comment]
5. How were the git submodules added to a specific commit or release tag?

**Reply by nmeum** [reply]
What do you mean?

**Reply by Biswa96** [reply]
How do you add git submodules in this project?

**Reply by nmeum** [reply]
The android repos have a tag for platform-tools releases (e.g. `platform-tools-31.0.3`). The submodules are pinned to this tag.

**Reply by Biswa96** [reply]
While exploring the submodules, I found that the commits does not match with latest release. For example, the last commit in adb in this repository dated 4th March. But [platform-tools-33.0.3](https://android.googlesource.com/platform/packages/modules/adb/+/refs/tags/platform-tools-33.0.3) tag has last commit in 16th August. Were submodules updated properly in last 33.0.3 update? Or am I missing something?

**Reply by nmeum** [reply]
Oh, indeed. Good catch! It looks like `vendor/adb` was last updated in commit 643b02609b16b9d3c804af871762bb95ebbb2fad for platform-tools-33.0.1. So it seems it was overlooked with the platform-tools-33.0.3 upgrade. Would probably be nice to have some sort of sanitycheck for this to prevent this from happening in future updates. Currently kind of difficult to manually keep track of all submodules during PR review.

**Reply by nmeum** [reply]
If you have the time, could you create an issue regarding the missing adb update so we don't forget about this? :)

**Reply by Biswa96** [reply]
I'll send a PR. Just CI testing is remaining.

**Comment by Biswa96** [comment]
6. Not a question, little update about upstreaming the patches from here. I have sent some patches from here to upstream. Some of the patches are in upstream now but some are not. I list some of the patches which are not accepted in upstream.
* 0004-Define-clang-only-nullability-specifiers-when-__clan.patch: https://android-review.googlesource.com/c/platform/packages/modules/adb/+/2361481

to be continued...

**Comment by Biswa96** [comment]
Inspired by this project design, I have built Android NDK toolchain, although it is not yet fully prepared. Could Linux distributions provide their own NDK toolchain compiled from source? Or, is the pre-built official Android NDK sufficient enough that Linux distributions do not need to build their own?
